### PR TITLE
Reduce ResponseTemplate cyclomatic complexity

### PR DESCRIPTION
**Pull request by on99**

### CODE REVIEW

File: design/dsl/api.go [hunk 1]
```diff
@@ -269,149 +270,64 @@ func ResponseTemplate(name string, p interface{}) {
...
+			return r
+		}
+		a.ResponseTemplates[name] = &design.ResponseTemplateDefinition{
+			Name:     name,
+			Template: t,
+		}
+	} else {
+		typ := reflect.TypeOf(p)
+		if kind := typ.Kind(); kind != reflect.Func {
+			ReportError("dsl should be a function but get %v", kind)
```
Comment (raphael) [inline design/dsl/api.go]: Please use:
`"dsl must be a function but got %v"`

File: design/dsl/api.go [hunk 2]
```diff
@@ -269,149 +270,64 @@ func ResponseTemplate(name string, p interface{}) {
...
+			return
+		}
+
+		if num := typ.NumOut(); num > 0 {
+			ReportError("dsl should have 0 return parameter but get %d", num)
```
Comment (raphael) [inline design/dsl/api.go]: Please use:
`"dsl functions cannot have return parameters"` (don't think we need to specify how many were actually set)

File: design/dsl/api.go [hunk 3]
```diff
@@ -269,149 +270,64 @@ func ResponseTemplate(name string, p interface{}) {
...
+			return
+		}
+
+		num := typ.NumIn()
+		val := reflect.ValueOf(p)
+		t := func(params ...string) *design.ResponseDefinition {
+			if len(params) < num {
+				ReportError("expected at least %d argument when invoking response template %s", num, name)
```
Comment (raphael) [inline design/dsl/api.go]: Can we have a `s

File: design/dsl/api.go [hunk 4]
```diff
@@ -269,149 +270,64 @@ func ResponseTemplate(name string, p interface{}) {
...
-			t := func(params ...string) *design.ResponseDefinition {
-				if len(params) < 9 {
-					ReportError("expected nine arguments when invoking response template %s", name)
+			r := &design.ResponseDefinition{Name: name}
+
+			in := make([]reflect.Value, num)
+			for i := 0; i < num; i++ {
+				// type checking
+				if t := typ.In(i); t.Kind() != reflect.String {
+					ReportError("expect parameter type string but get %s", t)
```
Comment (raphael) [inline design/dsl/api.go]: Please use:
`("ResponseTemplate parameters must be strings but type of parameter at position %d is %s", i, t)`

File: design/dsl/api.go (hunk 2 above)
Comment (on99) [inline design/dsl/api.go]: Yes we dont need to specify it but I just want to simulate your code.
Let me remove it if it makes more sense.

File: design/dsl/api.go (hunk 3 above)
Comment (on99) [inline design/dsl/api.go]: Will do.

File: design/dsl/api.go (hunk 4 above)
Comment (on99) [inline design/dsl/api.go]: Will do.

File: design/dsl/api.go (hunk 1 above)
Comment (on99) [inline design/dsl/api.go]: Will do.

**Comment by on99** [comment]
@raphael  updated, please check.

File: design/dsl/api.go (hunk 2 above)
Comment (raphael) [inline design/dsl/api.go]: I think it makes more sense to remove it, thanks for checking!

**Comment by raphael** [comment]
Thanks! please go ahead with comment 2 and I'll merge.

**Comment by gogames** [comment]
@raphael  I remove the code because I think it makes more sense that we do not check the return parameters.

**Comment by on99** [comment]
@raphael I remove the code because I think it makes more sense that we do not check the return parameters.

**Comment by raphael** [comment]
Makes sense. LGTM.

**Pull request merged**